src/rgw/rgw_rados.cc

```diff
@@ -6383,21 +6384,19 @@ int RGWRados::get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
     int r = append_atomic_test(astate, op);
     if (r < 0)
       return r;
-
-    if (astate &&
-        obj_ofs < astate->data.length()) {
-      unsigned chunk_len = std::min((uint64_t)astate->data.length() - obj_ofs, (uint64_t)len);
-
-      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+    if (astate && astate->prefetch
+      && astate->prefetch->off == obj_ofs) {
+      // diff is mainly for compression where the len could be smaller than data length
+      auto diff = std::min((unsigned int)len, astate->data.length());
+      r = d->client_cb->handle_data(astate->data, 0, diff);
```

Comment:
@fengchunsong I'm not sure I understood what you are suggesting, Can you please describe what is purpose of this change?
Do you have feature branch for it?

Thank you

Reply:
Check the call stack as follows. Because the offset is not 0, the system reads data in prefetch first and then reads data again. In this case, prefetch does not take effect.
Thread 357 "ganesha.nfsd" hit Breakpoint 1, Objecter::op_submit (this=0xffffaa484800, op=0xfffede974800, ptid=0x0, ctx_budget=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/osdc/Objecter.cc:2245
2245    {
$51 = {<RefCountedObject> = {_vptr.RefCountedObject = 0xffffb396b628 <vtable for Objecter::Op+16>, nref = {<std::__atomic_base<unsigned long>> = {static _S_alignment = 8, _M_i = 1}, static is_always_lock_free = true}, cct = 0x0}, session = 0x0, incarnation = 0, target = {flags = 16, epoch = 0, base_oid = {name = "6d612017-5874-4762-b677-49bdd3a090b2.217167.1_vdb.1_1.dir/vdb_f0008.file"}, base_oloc = {pool = 24, key = "", nspace = "", hash = -1}, target_oid = {name = ""}, target_oloc = {pool = -1, key = "", nspace = "", hash = -1}, precalc_pgid = false, pool_ever_existed = false, base_pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, actual_pgid = {pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, shard = {id = -1 '\377', static NO_SHARD = {id = -1 '\377', static NO_SHARD = <same as static member of an already seen type>}}, static calc_name_buf_size = 40 '('}, pg_num = 0, pg_num_mask = 0, pg_num_pending = 0, up = std::vector of length 0, capacity 0, acting = std::vector of length 0, capacity 0, up_primary = -1, acting_primary = -1, size = -1, min_size = -1, sort_bitwise = false, recovery_deletes = false, used_replica = false, paused = false, osd = -1, last_force_resend = 0}, con = {px = 0x0}, features = 4611087854035861503, ops = std::vector of length 3, capacity 4 = {{op = {op = {v = 4866}, flags = {v = 0}, {extent = {offset = {v = 0}, length = {v = 0}, truncate_size = {v = 0}, truncate_seq = {v = 0}}, xattr = {name_len = {v = 0}, value_len = {v = 0}, cmp_op = 0 '\000', cmp_mode = 0 '\000'}, cls = {class_len = 0 '\000', method_len = 0 '\000', argc = 0 '\000', indata_len = {v = 0}}, pgls = {count = {v = 0}, start_epoch = {v = 0}}, snap = {snapid = {v = 0}}, watch = {cookie = {v = 0}, ver = {v = 0}, op = 0 '\000', gen = 0, timeout = 0}, notify = {cookie = {v = 0}}, assert_ver = {unused = {v = 0}, ver = {v = 0}}, clonerange = {offset = {v = 0}, length = {v = 0}, src_offset = {v = 0}}, copy_get = {max = {v = 0}}, copy_from = {snapid = {v = 0}, src_version = {v = 0}, flags = 0 '\000', src_fadvise_flags = {v = 0}}, hit_set_get = {stamp = {tv_sec = {v = 0}, tv_nsec = {v = 0}}}, tmap2omap = {flags = 0 '\000'}, alloc_hint = {expected_object_size = {v = 0}, expected_write_size = {v = 0}, flags = {v = 0}}, writesame = {offset = {v = 0}, length = {v = 0}, data_length = {v = 0}}, checksum = {offset = {v = 0}, length = {v = 0}, chunk_size = {v = 0}, type = 0 '\000'}}, payload_len = {v = 0}}, soid = {oid = {name = ""}, snap = {val = 0}}, indata = {_buffers = {_root = {next = 0xfffede8e4050}, _tail = 0xfffede8e4050, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4050, ls = 0xfffede8e4050, p = {cur = 0xfffede8e4050}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, outdata = {_buffers = {_root = {next = 0xfffede8e4098}, _tail = 0xfffede8e4098, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4098, ls = 0xfffede8e4098, p = {cur = 0xfffede8e4098}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, rval = {code = 0}}, {op = {op = {v = 4610}, flags = {v = 0}, {extent = {offset = {v = 0}, length = {v = 0}, truncate_size = {v = 0}, truncate_seq = {v = 0}}, xattr = {name_len = {v = 0}, value_len = {v = 0}, cmp_op = 0 '\000', cmp_mode = 0 '\000'}, cls = {class_len = 0 '\000', method_len = 0 '\000', argc = 0 '\000', indata_len = {v = 0}}, pgls = {count = {v = 0}, start_epoch = {v = 0}}, snap = {snapid = {v = 0}}, watch = {cookie = {v = 0}, ver = {v = 0}, op = 0 '\000', gen = 0, timeout = 0}, notify = {cookie = {v = 0}}, assert_ver = {unused = {v = 0}, ver = {v = 0}}, clonerange = {offset = {v = 0}, length = {v = 0}, src_offset = {v = 0}}, copy_get = {max = {v = 0}}, copy_from = {snapid = {v = 0}, src_version = {v = 0}, flags = 0 '\000', src_fadvise_flags = {v = 0}}, hit_set_get = {stamp = {tv_sec = {v = 0}, tv_nsec = {v = 0}}}, tmap2omap = {flags = 0 '\000'}, alloc_hint = {expected_object_size = {v = 0}, expected_write_size = {v = 0}, flags = {v = 0}}, writesame = {offset = {v = 0}, length = {v = 0}, data_length = {v = 0}}, checksum = {offset = {v = 0}, length = {v = 0}, chunk_size = {v = 0}, type = 0 '\000'}}, payload_len = {v = 0}}, soid = {oid = {name = ""}, snap = {val = 0}}, indata = {_buffers = {_root = {next = 0xfffede8e4138}, _tail = 0xfffede8e4138, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4138, ls = 0xfffede8e4138, p = {cur = 0xfffede8e4138}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, outdata = {_buffers = {_root = {next = 0xfffede8e4180}, _tail = 0xfffede8e4180, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4180, ls = 0xfffede8e4180, p = {cur = 0xfffede8e4180}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, rval = {code = 0}}, {op = {op = {v = 4609}, flags = {v = 0}, {extent = {offset = {v = 0}, length = {v = 4194304}, truncate_size = {v = 0}, truncate_seq = {v = 0}}, xattr = {name_len = {v = 0}, value_len = {v = 0}, cmp_op = 0 '\000', cmp_mode = 0 '\000'}, cls = {class_len = 0 '\000', method_len = 0 '\000', argc = 0 '\000', indata_len = {v = 0}}, pgls = {count = {v = 0}, start_epoch = {v = 4194304}}, snap = {snapid = {v = 0}}, watch = {cookie = {v = 0}, ver = {v = 4194304}, op = 0 '\000', gen = 0, timeout = 0}, notify = {cookie = {v = 0}}, assert_ver = {unused = {v = 0}, ver = {v = 4194304}}, clonerange = {offset = {v = 0}, length = {v = 4194304}, src_offset = {v = 0}}, copy_get = {max = {v = 0}}, copy_from = {snapid = {v = 0}, src_version = {v = 4194304}, flags = 0 '\000', src_fadvise_flags = {v = 0}}, hit_set_get = {stamp = {tv_sec = {v = 0}, tv_nsec = {v = 0}}}, tmap2omap = {flags = 0 '\000'}, alloc_hint = {expected_object_size = {v = 0}, expected_write_size = {v = 4194304}, flags = {v = 0}}, writesame = {offset = {v = 0}, length = {v = 4194304}, data_length = {v = 0}}, checksum = {offset = {v = 0}, length = {v = 4194304}, chunk_size = {v = 0}, type = 0 '\000'}}, payload_len = {v = 0}}, soid = {oid = {name = ""}, snap = {val = 0}}, indata = {_buffers = {_root = {next = 0xfffede8e4220}, _tail = 0xfffede8e4220, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4220, ls = 0xfffede8e4220, p = {cur = 0xfffede8e4220}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, outdata = {_buffers = {_root = {next = 0xfffede8e4268}, _tail = 0xfffede8e4268, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede8e4268, ls = 0xfffede8e4268, p = {cur = 0xfffede8e4268}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, rval = {code = 0}}}, snapid = {val = 18446744073709551614}, snapc = {seq = {val = 0}, snaps = std::vector of length 0, capacity 0}, mtime = {__d = {__r = 0}}, outbl = 0xffff76bd7628, out_bl = std::vector of length 3, capacity 4 = {0xfffede7805c0, 0xfffedeed0688, 0xfffede3a23b8}, out_handler = std::vector of length 3, capacity 4 = {0xfffede7805b0, 0xfffedeed0680, 0x0}, out_rval = std::vector of length 3, capacity 4 = {0x0, 0x0, 0x0}, priority = 0, onfinish = 0xfffedefa0d50, ontimeout = 0, tid = 0, attempts = 0, objver = 0xffff76bd73f8, reply_epoch = 0x0, stamp = {__d = {__r = 0}}, map_dne_bound = 0, budget = -1, should_resend = true, ctx_budgeted = false, data_offset = 0x0, reqid = {name = {_type = 0 '\000', _num = 0, static TYPE_MON = 1, static TYPE_MDS = 2, static TYPE_OSD = 4, static TYPE_CLIENT = 8, static TYPE_MGR = 16, static NEW = -1}, tid = 0, inc = 0}, trace = {<No data fields>}}
#0  Objecter::op_submit (this=0xffffaa484800, op=0xfffede974800, ptid=0x0, ctx_budget=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/osdc/Objecter.cc:2245
#1  0x0000ffffb38adb1c in librados::IoCtxImpl::operate_read (this=0xfffedef31340, oid=..., o=0xfffedeed0600, pbl=pbl@entry=0xffff76bd7628, flags=flags@entry=0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/librados/IoCtxImpl.cc:725
#2  0x0000ffffb3884448 in librados::v14_2_0::IoCtx::operate (this=this@entry=0xffff76bd7710, oid="6d612017-5874-4762-b677-49bdd3a090b2.217167.1_vdb.1_1.dir/vdb_f0008.file", o=o@entry=0xffff76bd75e8, pbl=pbl@entry=0xffff76bd7628) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/librados/librados_cxx.cc:1423
#3  0x0000ffffa7ecd3bc in RGWRados::raw_obj_stat (this=this@entry=0xffffb3ca1800, obj=..., psize=psize@entry=0xfffede3a17f8, pmtime=pmtime@entry=0xfffede3a1808, epoch=epoch@entry=0xfffede3a1810, attrs=attrs@entry=0xfffede3a24c8, first_chunk=0xfffede3a23b8, prefetch=std::optional<prefetch_range> = {...}, objv_tracker=objv_tracker@entry=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:8007
#4  0x0000ffffa7eeba3c in RGWRados::get_obj_state_impl (this=this@entry=0xffffb3ca1800, rctx=rctx@entry=0xffff76bd9518, bucket_info=..., obj=..., state=state@entry=0xffff76bd7f00, follow_olh=follow_olh@entry=true, assume_noent=assume_noent@entry=false) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:5811
#5  0x0000ffffa7eed1b8 in RGWRados::get_obj_state (this=0xffffb3ca1800, rctx=0xffff76bd9518, bucket_info=..., obj=..., state=state@entry=0xffff76bd7f00, follow_olh=follow_olh@entry=true, assume_noent=assume_noent@entry=false) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:5968
#6  0x0000ffffa7eef52c in RGWRados::Object::get_state (this=<optimized out>, pstate=pstate@entry=0xffff76bd7f00, follow_olh=follow_olh@entry=true, assume_noent=assume_noent@entry=false) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:6114
#7  0x0000ffffa7eef5e4 in RGWRados::Object::Read::get_attr (this=this@entry=0xffff76bd8528, name=name@entry=0xffffa8155e90 "user.rgw.acl", dest=...) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:5990
#8  0x0000ffffa7e74004 in get_obj_policy_from_attr (bucket_attrs=std::map with 3 elements = {...}, obj=..., storage_class=0x0, policy=0xfffedc340280, bucket_info=..., obj_ctx=..., store=0xffffb3ca1800, cct=0xffffb3de0000) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_op.cc:247
#9  read_obj_policy (store=store@entry=0xffffb3ca1800, s=<optimized out>, s@entry=0xffff76bd95c0, bucket_info=..., bucket_attrs=std::map with 3 elements = {...}, acl=0xfffedc340280, storage_class=storage_class@entry=0x0, policy=..., bucket=..., object=...) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_op.cc:518
#10 0x0000ffffa7e74870 in rgw_build_object_policies (store=store@entry=0xffffb3ca1800, s=s@entry=0xffff76bd95c0, prefetch=0xffff76bdaab0) at /usr/include/c++/8/bits/unique_ptr.h:337
#11 0x0000ffffa7d2d984 in rgw::RGWLibRequest::read_permissions (this=0xffff76bda928, op=0xffff76bdaa18) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_lib.h:135
#12 0x0000ffffa7d2f300 in rgw::RGWLibProcess::process_request (this=this@entry=0xffffb3ca3800, req=req@entry=0xffff76bda928, io=io@entry=0xffff76bda5d0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/librgw.cc:279
#13 0x0000ffffa7d305d0 in rgw::RGWLibProcess::process_request (this=0xffffb3ca3800, req=req@entry=0xffff76bda928) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/librgw.cc:170
#14 0x0000ffffa7d3fb88 in rgw::RGWLibFrontend::execute_req (this=<optimized out>, req=0xffff76bda928) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_lib_frontend.h:98
#15 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=262144, length=1048576, bytes_read=0xffff76bdb0a0, buffer=0xfffeda287000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_file.cc:294
#16 0x0000ffffa8317e34 in rgw_fsal_read2 (obj_hdl=0xfffede940000, bypass=false, done_cb=0xffffb50f29c8 <mdc_read_cb>, read_arg=0xfffede9405b8, caller_arg=0xfffede800000) at /root/ganesha/nfs-ganesha/src/FSAL/FSAL_RGW/handle.c:1355
#17 0x0000ffffb50f2b50 in mdcache_read2 (obj_hdl=0xfffede882838, bypass=false, done_cb=0xffffb50b05e0 <nfs4_read_cb>, read_arg=0xfffede9405b8, caller_arg=0xfffede940540) at /root/ganesha/nfs-ganesha/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c:619
#18 0x0000ffffb50b12d8 in nfs4_read (op=0xfffedef312a0, data=0xfffedeea0180, resp=0xfffede8d2740, io=IO_READ, info=0x0) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_op_read.c:685
#19 0x0000ffffb50b13b4 in nfs4_op_read (op=0xfffedef312a0, data=0xfffedeea0180, resp=0xfffede8d2740) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_op_read.c:739
#20 0x0000ffffb50934d4 in process_one_op (data=0xfffedeea0180, status=0xffff76bdc38c) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_Compound.c:920
#21 0x0000ffffb50943b8 in nfs4_Compound (arg=0xfffedeee0820, req=0xfffedeee0000, res=0xfffedee50a00) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_Compound.c:1324
#22 0x0000ffffb4fe5c04 in nfs_rpc_process_request (reqdata=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1500
#23 0x0000ffffb4fe5f54 in nfs_rpc_valid_NFS (req=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1629
#24 0x0000ffffb4f40380 in svc_vc_decode (req=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_vc.c:904
#25 0x0000ffffb4f3bf2c in svc_request (xprt=0xffff656a0000, xdrs=0xfffedeea0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1202
#26 0x0000ffffb4f40280 in svc_vc_recv (xprt=0xffff656a0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_vc.c:877
#27 0x0000ffffb4f3beac in svc_rqst_xprt_task_recv (wpe=0xffff656a0240) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1183
#28 0x0000ffffb4f3cb44 in svc_rqst_epoll_loop (wpe=0xffffaa485b20) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1567
#29 0x0000ffffb4f48100 in work_pool_thread (arg=0xffff964923a0) at /root/ganesha/nfs-ganesha/src/libntirpc/src/work_pool.c:183
#30 0x0000ffffb4a877f8 in start_thread () from /lib64/libpthread.so.0
#31 0x0000ffffb4930edc in thread_start () from /lib64/libc.so.6

then read again in

Thread 357 "ganesha.nfsd" hit Breakpoint 1, Objecter::op_submit (this=0xffffaa484800, op=0xfffede975100, ptid=0xfffedeea1a40, ctx_budget=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/osdc/Objecter.cc:2245
2245    {
$54 = {<RefCountedObject> = {_vptr.RefCountedObject = 0xffffb396b628 <vtable for Objecter::Op+16>, nref = {<std::__atomic_base<unsigned long>> = {static _S_alignment = 8, _M_i = 1}, static is_always_lock_free = true}, cct = 0x0}, session = 0x0, incarnation = 0, target = {flags = 32, epoch = 0, base_oid = {name = "2020-11-03-16-6d612017-5874-4762-b677-49bdd3a090b2.217167.1-rgwtestc4me"}, base_oloc = {pool = 27, key = "", nspace = "", hash = -1}, target_oid = {name = ""}, target_oloc = {pool = -1, key = "", nspace = "", hash = -1}, precalc_pgid = false, pool_ever_existed = false, base_pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, actual_pgid = {pgid = {m_pool = 0, m_seed = 0, static calc_name_buf_size = 36 '$'}, shard = {id = -1 '\377', static NO_SHARD = {id = -1 '\377', static NO_SHARD = <same as static member of an already seen type>}}, static calc_name_buf_size = 40 '('}, pg_num = 0, pg_num_mask = 0, pg_num_pending = 0, up = std::vector of length 0, capacity 0, acting = std::vector of length 0, capacity 0, up_primary = -1, acting_primary = -1, size = -1, min_size = -1, sort_bitwise = false, recovery_deletes = false, used_replica = false, paused = false, osd = -1, last_force_resend = 0}, con = {px = 0x0}, features = 4611087854035861503, ops = std::vector of length 1, capacity 1 = {{op = {op = {v = 8710}, flags = {v = 0}, {extent = {offset = {v = 0}, length = {v = 359}, truncate_size = {v = 0}, truncate_seq = {v = 0}}, xattr = {name_len = {v = 0}, value_len = {v = 0}, cmp_op = 103 'g', cmp_mode = 1 '\001'}, cls = {class_len = 0 '\000', method_len = 0 '\000', argc = 0 '\000', indata_len = {v = 0}}, pgls = {count = {v = 0}, start_epoch = {v = 359}}, snap = {snapid = {v = 0}}, watch = {cookie = {v = 0}, ver = {v = 359}, op = 0 '\000', gen = 0, timeout = 0}, notify = {cookie = {v = 0}}, assert_ver = {unused = {v = 0}, ver = {v = 359}}, clonerange = {offset = {v = 0}, length = {v = 359}, src_offset = {v = 0}}, copy_get = {max = {v = 0}}, copy_from = {snapid = {v = 0}, src_version = {v = 359}, flags = 0 '\000', src_fadvise_flags = {v = 0}}, hit_set_get = {stamp = {tv_sec = {v = 0}, tv_nsec = {v = 0}}}, tmap2omap = {flags = 0 '\000'}, alloc_hint = {expected_object_size = {v = 0}, expected_write_size = {v = 359}, flags = {v = 0}}, writesame = {offset = {v = 0}, length = {v = 359}, data_length = {v = 0}}, checksum = {offset = {v = 0}, length = {v = 359}, chunk_size = {v = 0}, type = 0 '\000'}}, payload_len = {v = 0}}, soid = {oid = {name = ""}, snap = {val = 0}}, indata = {_buffers = {_root = {next = 0xfffee0211e00}, _tail = 0xfffee0211e00, _size = 1}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 359, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede2a0550, ls = 0xfffede2a0550, p = {cur = 0xfffee0211e00}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, outdata = {_buffers = {_root = {next = 0xfffede2a0598}, _tail = 0xfffede2a0598, _size = 0}, _carriage = 0xffffb39765b0 <ceph::buffer::v14_2_0::list::always_empty_bptr>, _len = 0, _memcopy_count = 0, last_p = {<ceph::buffer::v14_2_0::list::iterator_impl<false>> = {bl = 0xfffede2a0598, ls = 0xfffede2a0598, p = {cur = 0xfffede2a0598}, off = 0, p_off = 0}, <No data fields>}, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}, static CLAIM_DEFAULT = 0, static CLAIM_ALLOW_NONSHAREABLE = 1}, rval = {code = 0}}}, snapid = {val = 18446744073709551614}, snapc = {seq = {val = 0}, snaps = std::vector of length 0, capacity 0}, mtime = {__d = {__r = 1604391778073019542}}, outbl = 0x0, out_bl = std::vector of length 1, capacity 1 = {0x0}, out_handler = std::vector of length 1, capacity 1 = {0x0}, out_rval = std::vector of length 1, capacity 1 = {0x0}, priority = 0, onfinish = 0xfffede810560, ontimeout = 0, tid = 0, attempts = 0, objver = 0xfffedeea1a38, reply_epoch = 0x0, stamp = {__d = {__r = 0}}, map_dne_bound = 0, budget = -1, should_resend = true, ctx_budgeted = false, data_offset = 0x0, reqid = {name = {_type = 0 '\000', _num = 0, static TYPE_MON = 1, static TYPE_MDS = 2, static TYPE_OSD = 4, static TYPE_CLIENT = 8, static TYPE_MGR = 16, static NEW = -1}, tid = 0, inc = 0}, trace = {<No data fields>}}
#0  Objecter::op_submit (this=0xffffaa484800, op=0xfffede975100, ptid=0xfffedeea1a40, ctx_budget=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/osdc/Objecter.cc:2245
#1  0x0000ffffb38a9be4 in librados::IoCtxImpl::aio_append (this=this@entry=0xfffedef31340, oid=..., c=0xfffedeea1980, bl=..., len=len@entry=359) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/librados/IoCtxImpl.cc:1018
#2  0x0000ffffb3885694 in librados::v14_2_0::IoCtx::aio_append (this=this@entry=0xffff76bd8c20, oid="2020-11-03-16-6d612017-5874-4762-b677-49bdd3a090b2.217167.1-rgwtestc4me", c=c@entry=0xfffedefd0058, bl=..., len=len@entry=359) at /usr/include/c++/8/bits/basic_string.h:936
#3  0x0000ffffa7eca9e8 in RGWRados::append_async (this=this@entry=0xffffb3ca1800, obj=..., size=359, bl=...) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_rados.cc:8588
#4  0x0000ffffa7d684c0 in rgw_log_op (store=store@entry=0xffffb3ca1800, rest=rest@entry=0x0, s=s@entry=0xffff76bd95c0, op_name="get_obj", olog=0x0) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/include/buffer.h:1050
#5  0x0000ffffa7d2e9c8 in rgw::RGWLibProcess::process_request (this=this@entry=0xffffb3ca3800, req=req@entry=0xffff76bda928, io=io@entry=0xffff76bda5d0) at /usr/include/c++/8/ext/new_allocator.h:79
#6  0x0000ffffa7d305d0 in rgw::RGWLibProcess::process_request (this=0xffffb3ca3800, req=req@entry=0xffff76bda928) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/librgw.cc:170
#7  0x0000ffffa7d3fb88 in rgw::RGWLibFrontend::execute_req (this=<optimized out>, req=0xffff76bda928) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_lib_frontend.h:98
#8  rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=262144, length=1048576, bytes_read=0xffff76bdb0a0, buffer=0xfffeda287000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-2-g402bd25a58/src/rgw/rgw_file.cc:294
#9  0x0000ffffa8317e34 in rgw_fsal_read2 (obj_hdl=0xfffede940000, bypass=false, done_cb=0xffffb50f29c8 <mdc_read_cb>, read_arg=0xfffede9405b8, caller_arg=0xfffede800000) at /root/ganesha/nfs-ganesha/src/FSAL/FSAL_RGW/handle.c:1355
#10 0x0000ffffb50f2b50 in mdcache_read2 (obj_hdl=0xfffede882838, bypass=false, done_cb=0xffffb50b05e0 <nfs4_read_cb>, read_arg=0xfffede9405b8, caller_arg=0xfffede940540) at /root/ganesha/nfs-ganesha/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c:619
#11 0x0000ffffb50b12d8 in nfs4_read (op=0xfffedef312a0, data=0xfffedeea0180, resp=0xfffede8d2740, io=IO_READ, info=0x0) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_op_read.c:685
#12 0x0000ffffb50b13b4 in nfs4_op_read (op=0xfffedef312a0, data=0xfffedeea0180, resp=0xfffede8d2740) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_op_read.c:739
#13 0x0000ffffb50934d4 in process_one_op (data=0xfffedeea0180, status=0xffff76bdc38c) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_Compound.c:920
#14 0x0000ffffb50943b8 in nfs4_Compound (arg=0xfffedeee0820, req=0xfffedeee0000, res=0xfffedee50a00) at /root/ganesha/nfs-ganesha/src/Protocols/NFS/nfs4_Compound.c:1324
#15 0x0000ffffb4fe5c04 in nfs_rpc_process_request (reqdata=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1500
#16 0x0000ffffb4fe5f54 in nfs_rpc_valid_NFS (req=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1629
#17 0x0000ffffb4f40380 in svc_vc_decode (req=0xfffedeee0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_vc.c:904
#18 0x0000ffffb4f3bf2c in svc_request (xprt=0xffff656a0000, xdrs=0xfffedeea0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1202
#19 0x0000ffffb4f40280 in svc_vc_recv (xprt=0xffff656a0000) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_vc.c:877
#20 0x0000ffffb4f3beac in svc_rqst_xprt_task_recv (wpe=0xffff656a0240) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1183
#21 0x0000ffffb4f3cb44 in svc_rqst_epoll_loop (wpe=0xffffaa485b20) at /root/ganesha/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1567
#22 0x0000ffffb4f48100 in work_pool_thread (arg=0xffff964923a0) at /root/ganesha/nfs-ganesha/src/libntirpc/src/work_pool.c:183
#23 0x0000ffffb4a877f8 in start_thread () from /lib64/libpthread.so.0
#24 0x0000ffffb4930edc in thread_start () from /lib64/libc.so.6
[Switching to Thread 0xffff79decf60 (LWP 243070)]

Reply:
use this patch:
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=0, length=262144, bytes_read=0xffff79deb0a0, buffer=0xfffedc188000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=0, length=262144, bytes_read=0xffff79deb0a0, buffer=0xfffedc188000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=262144, length=1048576, bytes_read=0xffff76bdb0a0, buffer=0xfffeda287000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=1310720, length=1048576, bytes_read=0xffff79deb0a0, buffer=0xfffedbee1000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=2359296, length=786432, bytes_read=0xffff781eb0a0, buffer=0xfffedc97f000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=262144, length=1048576, bytes_read=0xffff76bdb0a0, buffer=0xfffeda287000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=1310720, length=1048576, bytes_read=0xffff79deb0a0, buffer=0xfffedbee1000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffedee50c80, offset=2359296, length=786432, bytes_read=0xffff781eb0a0, buffer=0xfffedc97f000, flags=<optimized out>)

modify your patch
if (astate && astate->prefetch
      && astate->prefetch->off == obj_ofs) 
to
off_t end = astate->prefetch->off + astate->data.length();	  
if (astate && astate->prefetch
&& (off_t)astate->prefetch->off <= obj_ofs && obj_ofs < end) {

If the offset is within the prefetch range, it is not read again.

rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffea7d40000, offset=0, length=262144, bytes_read=0xffff467eb0a0, buffer=0xfffea781d000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffea7d40000, offset=262144, length=1048576, bytes_read=0xffff451db0a0, buffer=0xfffea6482000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffea7d40000, offset=1310720, length=1048576, bytes_read=0xffff467eb0a0, buffer=0xfffea7869000, flags=<optimized out>)
rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffea7d40000, offset=2359296, length=786432, bytes_read=0xffff4750b0a0, buffer=0xfffea57b2000, flags=<optimized out>)

Reply:
@fengchunsong can you verify if adding those lines solving the issue?

+++ b/src/rgw/rgw_file.h
@@ -1976,6 +1976,8 @@ public:
     RGWGetObj::partial_content = true;
     RGWGetObj::ofs = off;
     RGWGetObj::end = off + len;
**+    prefetch.off = off;
+    prefetch.len = len;**
   }

Reply:
Looks good. I'll verify it right away.

Reply:
I add this two lines,it still have prefetch read first and then read again

Reply:
@fengchunsong I added a new commit that fixes it, can you try again and let me know?

Reply:
I test read a 3MB object with this rgw: Fix librgw end define patch,According to the logs, there are still two ops.
	Line 377: #15 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=0, length=262144, bytes_read=0xffff7033c910, buffer=0xfffe747f5000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 419: #15 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=1310720, length=1048576, bytes_read=0xffff7033c910, buffer=0xfffe74c8b000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 463: #15 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=262144, length=1048576, bytes_read=0xffff7044c910, buffer=0xfffdc47f2000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 506: #15 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=2359296, length=786432, bytes_read=0xffff583dc910, buffer=0xfffdbc003000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 546: #12 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=1310720, length=1048576, bytes_read=0xffff7033c910, buffer=0xfffe74c8b000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 587: #12 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=262144, length=1048576, bytes_read=0xffff7044c910, buffer=0xfffdc47f2000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294
	Line 627: #12 rgw::RGWLibFS::read (this=<optimized out>, rgw_fh=0xfffe747ef350, offset=2359296, length=786432, bytes_read=0xffff583dc910, buffer=0xfffdbc003000, flags=<optimized out>) at /home/rpmbuild/BUILD/ceph-14.2.12-3-gbda28912cf/src/rgw/rgw_file.cc:294

Reply:
I uploaded the call stack log to the tracker. Please check at https://tracker.ceph.com/issues/48241